src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
@poettering  What do you think about this?
> Is it worth the complexity to have one file per file kind? What benefit does it bring? A device with 0:0 returning ENODEV on any operation looked good enough for me.

Also for the systemd-nspawn extension how should we proceed? [DeviceAllow=](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#DeviceAllow=) does not support major/minor passing. How can we introduce this capability?